### PR TITLE
🎁 IOTSTOOL1176 - websockets

### PR DESCRIPTION
**Pull request by alexl0gan**

### CODE REVIEW

File: src/mbed_cloud/connect/notifications.py [hunk 1]
```diff
@@ -200,26 +212,151 @@ def __init__(self, db, queues, b64decode=True, notifications_api=None,
...
+        return on_message
+
+    def _get_on_close_callback(self):
+        def on_close(ws, error, reason):
+            if error == 1000:
+                self._delete_websocket_channel()
+            elif error == 1008:
+                self._log_error(error, reason)
+            elif error == 1006:
+                self._start_websocket()
```
Comment (madchutney) [inline src/mbed_cloud/connect/notifications.py]: I think calling the start websocket method which calls the run_forever is eventually going to blow the stack as it is recursive and will recurse every time the socket closes.

File: src/mbed_cloud/connect/notifications.py [hunk 2]
```diff
@@ -200,26 +212,151 @@ def __init__(self, db, queues, b64decode=True, notifications_api=None,
...
+                queues=self.queues,
+                b64decode=self._b64decode,
+                notification_object=data
+            )
+            if self.subscription_manager:
+                self.subscription_manager.notify(data.to_dict())
+
+        return on_message
+
+    def _get_on_close_callback(self):
```
[the remark this answers is not in the document]
Comment (madchutney) [reply src/mbed_cloud/connect/notifications.py]: Discussed this and the answer is _no_ as access to `self` is required.

**Comment by alexl0gan** [comment]
Ready for second look @madchutney @acabarbaye

**Comment by madchutney** [comment]
This is functionally complete but cannot be merged until the WebSocket is available in production.

**Pull request merged**